@WalterBright FreeBSD_32 actually [fails](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=2458887&isPull=true) with a linker error:
```
../dmd/src/dmd -conf= -I../druntime/import  -w -dip25 -m32 -fPIC -O -release -shared -unittest -ofgenerated/freebsd/release/32/unittest/libphobos2-ut.so generated/freebsd/release/32/unittest/std/array.o generated/freebsd/release/32/unittest/std/ascii.o generated/freebsd/release/32/unittest/std/base64.o generated/freebsd/release/32/unittest/std/bigint.o generated/freebsd/release/32/unittest/std/bitmanip.o generated/freebsd/release/32/unittest/std/compiler.o generated/freebsd/release/32/unittest/std/complex.o generated/freebsd/release/32/unittest/std/concurrency.o generated/freebsd/release/32/unittest/std/conv.o generated/freebsd/release/32/unittest/std/csv.o generated/freebsd/release/32/unittest/std/datetime.o generated/freebsd/release/32/unittest/std/demangle.o generated/freebsd/release/32/unittest/std/encoding.o generated/freebsd/release/32/unittest/std/exception.o generated/freebsd/release/32/unittest/std/file.o generated/freebsd/release/32/unittest/std/format.o generated/freebsd/release/32/unittest/std/functional.o generated/freebsd/release/32/unittest/std/getopt.o generated/freebsd/release/32/unittest/std/json.o generated/freebsd/release/32/unittest/std/math.o generated/freebsd/release/32/unittest/std/mathspecial.o generated/freebsd/release/32/unittest/std/meta.o generated/freebsd/release/32/unittest/std/mmfile.o generated/freebsd/release/32/unittest/std/numeric.o generated/freebsd/release/32/unittest/std/outbuffer.o generated/freebsd/release/32/unittest/std/parallelism.o generated/freebsd/release/32/unittest/std/path.o generated/freebsd/release/32/unittest/std/process.o generated/freebsd/release/32/unittest/std/random.o generated/freebsd/release/32/unittest/std/signals.o generated/freebsd/release/32/unittest/std/socket.o generated/freebsd/release/32/unittest/std/stdint.o generated/freebsd/release/32/unittest/std/stdio.o generated/freebsd/release/32/unittest/std/string.o generated/freebsd/release/32/unittest/std/system.o generated/freebsd/release/32/unittest/std/traits.o generated/freebsd/release/32/unittest/std/typecons.o generated/freebsd/release/32/unittest/std/typetuple.o generated/freebsd/release/32/unittest/std/uni.o generated/freebsd/release/32/unittest/std/uri.o generated/freebsd/release/32/unittest/std/utf.o generated/freebsd/release/32/unittest/std/uuid.o generated/freebsd/release/32/unittest/std/variant.o generated/freebsd/release/32/unittest/std/xml.o generated/freebsd/release/32/unittest/std/zip.o generated/freebsd/release/32/unittest/std/zlib.o generated/freebsd/release/32/unittest/std/algorithm/comparison.o generated/freebsd/release/32/unittest/std/algorithm/iteration.o generated/freebsd/release/32/unittest/std/algorithm/mutation.o generated/freebsd/release/32/unittest/std/algorithm/package.o generated/freebsd/release/32/unittest/std/algorithm/searching.o generated/freebsd/release/32/unittest/std/algorithm/setops.o generated/freebsd/release/32/unittest/std/algorithm/sorting.o generated/freebsd/release/32/unittest/std/container/array.o generated/freebsd/release/32/unittest/std/container/binaryheap.o generated/freebsd/release/32/unittest/std/container/dlist.o generated/freebsd/release/32/unittest/std/container/package.o generated/freebsd/release/32/unittest/std/container/rbtree.o generated/freebsd/release/32/unittest/std/container/slist.o generated/freebsd/release/32/unittest/std/container/util.o generated/freebsd/release/32/unittest/std/digest/crc.o generated/freebsd/release/32/unittest/std/digest/digest.o generated/freebsd/release/32/unittest/std/digest/hmac.o generated/freebsd/release/32/unittest/std/digest/md.o generated/freebsd/release/32/unittest/std/digest/murmurhash.o generated/freebsd/release/32/unittest/std/digest/ripemd.o generated/freebsd/release/32/unittest/std/digest/sha.o generated/freebsd/release/32/unittest/std/experimental/allocator/common.o generated/freebsd/release/32/unittest/std/experimental/allocator/gc_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/mallocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/mmap_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/package.o generated/freebsd/release/32/unittest/std/experimental/allocator/showcase.o generated/freebsd/release/32/unittest/std/experimental/allocator/typed.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/affix_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/allocator_list.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/bucketizer.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/fallback_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/free_list.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/free_tree.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/bitmapped_block.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/kernighan_ritchie.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/null_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/package.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/quantizer.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/region.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/scoped_allocator.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/segregator.o generated/freebsd/release/32/unittest/std/experimental/allocator/building_blocks/stats_collector.o generated/freebsd/release/32/unittest/std/experimental/logger/core.o generated/freebsd/release/32/unittest/std/experimental/logger/filelogger.o generated/freebsd/release/32/unittest/std/experimental/logger/nulllogger.o generated/freebsd/release/32/unittest/std/experimental/logger/multilogger.o generated/freebsd/release/32/unittest/std/experimental/logger/package.o generated/freebsd/release/32/unittest/std/net/curl.o generated/freebsd/release/32/unittest/std/net/isemail.o generated/freebsd/release/32/unittest/std/experimental/checkedint.o generated/freebsd/release/32/unittest/std/experimental/typecons.o generated/freebsd/release/32/unittest/std/range/interfaces.o generated/freebsd/release/32/unittest/std/range/package.o generated/freebsd/release/32/unittest/std/range/primitives.o generated/freebsd/release/32/unittest/std/regex/package.o generated/freebsd/release/32/unittest/std/regex/internal/generator.o generated/freebsd/release/32/unittest/std/regex/internal/ir.o generated/freebsd/release/32/unittest/std/regex/internal/parser.o generated/freebsd/release/32/unittest/std/regex/internal/backtracking.o generated/freebsd/release/32/unittest/std/regex/internal/tests.o generated/freebsd/release/32/unittest/std/regex/internal/thompson.o generated/freebsd/release/32/unittest/std/regex/internal/kickstart.o generated/freebsd/release/32/unittest/std/c/linux/linux.o generated/freebsd/release/32/unittest/std/c/linux/socket.o generated/freebsd/release/32/unittest/std/c/windows/com.o generated/freebsd/release/32/unittest/std/c/windows/stat.o generated/freebsd/release/32/unittest/std/c/windows/windows.o generated/freebsd/release/32/unittest/std/c/windows/winsock.o generated/freebsd/release/32/unittest/std/windows/charset.o generated/freebsd/release/32/unittest/std/windows/iunknown.o generated/freebsd/release/32/unittest/std/windows/syserror.o generated/freebsd/release/32/unittest/etc/c/curl.o generated/freebsd/release/32/unittest/etc/c/odbc/sql.o generated/freebsd/release/32/unittest/etc/c/odbc/sqlext.o generated/freebsd/release/32/unittest/etc/c/odbc/sqltypes.o generated/freebsd/release/32/unittest/etc/c/odbc/sqlucode.o generated/freebsd/release/32/unittest/etc/c/sqlite3.o generated/freebsd/release/32/unittest/etc/c/zlib.o generated/freebsd/release/32/unittest/std/c/fenv.o generated/freebsd/release/32/unittest/std/c/locale.o generated/freebsd/release/32/unittest/std/c/math.o generated/freebsd/release/32/unittest/std/c/process.o generated/freebsd/release/32/unittest/std/c/stdarg.o generated/freebsd/release/32/unittest/std/c/stddef.o generated/freebsd/release/32/unittest/std/c/stdio.o generated/freebsd/release/32/unittest/std/c/stdlib.o generated/freebsd/release/32/unittest/std/c/string.o generated/freebsd/release/32/unittest/std/c/time.o generated/freebsd/release/32/unittest/std/c/wcharh.o generated/freebsd/release/32/unittest/std/algorithm/internal.o generated/freebsd/release/32/unittest/std/concurrencybase.o generated/freebsd/release/32/unittest/std/internal/cstring.o generated/freebsd/release/32/unittest/std/internal/digest/sha_SSSE3.o generated/freebsd/release/32/unittest/std/internal/encodinginit.o generated/freebsd/release/32/unittest/std/internal/math/biguintcore.o generated/freebsd/release/32/unittest/std/internal/math/biguintnoasm.o generated/freebsd/release/32/unittest/std/internal/math/biguintx86.o generated/freebsd/release/32/unittest/std/internal/math/errorfunction.o generated/freebsd/release/32/unittest/std/internal/math/gammafunction.o generated/freebsd/release/32/unittest/std/internal/processinit.o generated/freebsd/release/32/unittest/std/internal/scopebuffer.o generated/freebsd/release/32/unittest/std/internal/test/dummyrange.o generated/freebsd/release/32/unittest/std/internal/unicode_comp.o generated/freebsd/release/32/unittest/std/internal/unicode_decomp.o generated/freebsd/release/32/unittest/std/internal/unicode_grapheme.o generated/freebsd/release/32/unittest/std/internal/unicode_norm.o generated/freebsd/release/32/unittest/std/internal/unicode_tables.o generated/freebsd/release/32/unittest/std/stdiobase.o generated/freebsd/release/32/etc/c/zlib/adler32.o generated/freebsd/release/32/etc/c/zlib/compress.o generated/freebsd/release/32/etc/c/zlib/crc32.o generated/freebsd/release/32/etc/c/zlib/deflate.o generated/freebsd/release/32/etc/c/zlib/gzclose.o generated/freebsd/release/32/etc/c/zlib/gzlib.o generated/freebsd/release/32/etc/c/zlib/gzread.o generated/freebsd/release/32/etc/c/zlib/gzwrite.o generated/freebsd/release/32/etc/c/zlib/infback.o generated/freebsd/release/32/etc/c/zlib/inffast.o generated/freebsd/release/32/etc/c/zlib/inflate.o generated/freebsd/release/32/etc/c/zlib/inftrees.o generated/freebsd/release/32/etc/c/zlib/trees.o generated/freebsd/release/32/etc/c/zlib/uncompr.o generated/freebsd/release/32/etc/c/zlib/zutil.o ../druntime/generated/freebsd/release/32/libdruntime.so.a  -defaultlib= -debuglib=
/usr/bin/ld: BFD 2.15 [FreeBSD] 2004-05-23 internal error, aborting at /usr/src/gnu/usr.bin/binutils/libbfd/../../../../contrib/binutils/bfd/elfcode.h line 213 in bfd_elf32_swap_symbol_out

/usr/bin/ld: Please report this bug.

Error: linker exited with status 1
posix.mak:349: recipe for target 'generated/freebsd/debug/32/unittest/libphobos2-ut.so' failed
gmake[1]: *** [generated/freebsd/debug/32/unittest/libphobos2-ut.so] Error 1
gmake[1]: Leaving directory '/home/braddr/sandbox/at-client/pull-2458887-FreeBSD_32/phobos'
posix.mak:275: recipe for target 'unittest-debug' failed
gmake: *** [unittest-debug] Error 2
gmake: *** Waiting for unfinished jobs....
/usr/bin/ld: BFD 2.15 [FreeBSD] 2004-05-23 internal error, aborting at /usr/src/gnu/usr.bin/binutils/libbfd/../../../../contrib/binutils/bfd/elfcode.h line 213 in bfd_elf32_swap_symbol_out

/usr/bin/ld: Please report this bug.

Error: linker exited with status 1
posix.mak:349: recipe for target 'generated/freebsd/release/32/unittest/libphobos2-ut.so' failed
gmake[1]: *** [generated/freebsd/release/32/unittest/libphobos2-ut.so] Error 1
gmake[1]: Leaving directory '/home/braddr/sandbox/at-client/pull-2458887-FreeBSD_32/phobos'
posix.mak:275: recipe for target 'unittest-release' failed
gmake: *** [unittest-release] Error 2
```

--- a/src/ddmd/backend/dwarf.c
+++ b/src/ddmd/backend/dwarf.c
@@ -73,7 +73,9 @@ unsigned CIE_offset_no_unwind;  // CIE offset for no unwind data
 IDXSYM elf_addsym(IDXSTR nam, targ_size_t val, unsigned sz,
         unsigned typ, unsigned bind, IDXSEC sec,
         unsigned char visibility = STV_DEFAULT);
+void addSegmentToComdat(segidx_t seg, segidx_t comdatseg);
 #endif
+
 
 static Outbuffer  *reset_symbuf;        // Keep pointers to reset symbols
 
@@ -117,10 +119,10 @@ int dwarf_getsegment(const char *name, int align)
 #endif
 }
 
-int dwarf_getsegment_alloc(const char *name, int align)
+int dwarf_getsegment_alloc(const char *name, const char *suffix, int align)
 {
 #if ELFOBJ
-    return ElfObj::getsegment(name, NULL, SHT_PROGBITS, SHF_ALLOC, align * 4);
+    return ElfObj::getsegment(name, suffix, SHT_PROGBITS, SHF_ALLOC, align * 4);
 #elif MACHOBJ
     return MachObj::getsegment(name, "__DWARF", align * 2, S_ATTR_DEBUG);
 #else
@@ -129,10 +131,24 @@ int dwarf_getsegment_alloc(const char *name, int align)
 #endif
 }
 
-int dwarf_except_table_alloc()
+int dwarf_except_table_alloc(Symbol *s)
 {
+    //printf("dwarf_except_table_alloc('%s')\n", s->Sident);
 #if ELFOBJ
-    return dwarf_getsegment_alloc(".gcc_except_table", 1);
+    /* If `s` is in a COMDAT, then this table needs to go into
+     * a unique section, which then gets added to the COMDAT group
+     * associated with `s`.
+     */
+    seg_data *pseg = SegData[s->Sseg];
+    if (pseg->SDassocseg)
+    {
+        const char *suffix = s->Sident; // cpp_mangle(s);
+        segidx_t tableseg = ElfObj::getsegment(".gcc_except_table.", suffix, SHT_PROGBITS, SHF_ALLOC|SHF_GROUP, 1);
+        addSegmentToComdat(tableseg, s->Sseg);
+        return tableseg;
+    }
+    else
+        return dwarf_getsegment_alloc(".gcc_except_table", NULL, 1);
 #elif MACHOBJ
     int seg = MachObj::getsegment("__gcc_except_tab", "__TEXT", 2, S_REGULAR);
     except_table_seg = seg;
@@ -146,7 +162,7 @@ int dwarf_except_table_alloc()
 int dwarf_eh_frame_alloc()
 {
 #if ELFOBJ
-    return dwarf_getsegment_alloc(".eh_frame", I64 ? 2 : 1);
+    return dwarf_getsegment_alloc(".eh_frame", NULL, I64 ? 2 : 1);
 #elif MACHOBJ
     int seg = MachObj::getsegment("__eh_frame", "__TEXT", I64 ? 3 : 2,
         S_COALESCED | S_ATTR_NO_TOC | S_ATTR_STRIP_STATIC_SYMS | S_ATTR_LIVE_SUPPORT);
@@ -951,7 +967,7 @@ void writeEhFrameFDE(IDXSEC dfseg, Symbol *sfunc, bool ehunwind, unsigned CIE_of
 #endif
     if (ehunwind)
     {
-        int etseg = dwarf_except_table_alloc();
+        int etseg = dwarf_except_table_alloc(sfunc);
 #if ELFOBJ
         buf->writeByten(4);                             // Augmentation Data Length
         buf->write32(I64 ? 0 : sfunc->Sfunc->LSDAoffset); // address of LSDA (".gcc_except_table")
@@ -2943,7 +2959,7 @@ void dwarf_except_gentables(Funcsym *sfunc, unsigned startoffset, unsigned retof
     if (!doUnwindEhFrame())
         return;
 
-    int seg = dwarf_except_table_alloc();
+    int seg = dwarf_except_table_alloc(sfunc);
     Outbuffer *buf = SegData[seg]->SDbuf;
     buf->reserve(100);
 

--- a/src/ddmd/backend/elfobj.c
+++ b/src/ddmd/backend/elfobj.c
@@ -75,11 +75,23 @@ char *obj_mangle2(Symbol *s,char *dest, size_t *destlen);
 #define cpp_mangle(s) ((s)->Sident)
 #endif
 
+void addSegmentToComdat(segidx_t seg, segidx_t comdatseg);
+
 /**
  * If set the compiler requires full druntime support of the new
  * section registration.
  */
 #define REQUIRE_DSO_REGISTRY (DMDV2 && (TARGET_LINUX || TARGET_FREEBSD))
+
+/******
+ * FreeBSD uses ELF, but the linker crashes with Elf comdats with the following message:
+ *  /usr/bin/ld: BFD 2.15 [FreeBSD] 2004-05-23 internal error, aborting at
+ *  /usr/src/gnu/usr.bin/binutils/libbfd/../../../../contrib/binutils/bfd/elfcode.h
+ *  line 213 in bfd_elf32_swap_symbol_out
+ * For the time being, just stick with Linux.
+ */
+
+#define ELF_COMDAT      TARGET_LINUX
 
 /***************************************************
  * Correspondence of relocation types
@@ -127,6 +139,7 @@ STATIC void obj_tlssections();
 #if MARS
 static void obj_rtinit();
 #endif
+static void addSectionToComdat(IDXSEC secidx, segidx_t comdatseg);
 
 static IDXSYM elf_addsym(IDXSTR sym, targ_size_t val, unsigned sz,
                          unsigned typ,unsigned bind,IDXSEC sec,
@@ -151,6 +164,8 @@ static Outbuffer *section_names;
 
 // Hash table for section_names
 AArray *section_names_hashtable;
+
+int jmpseg;
 
 /* ====================== Cached Strings in section_names ================= */
 
@@ -1677,6 +1692,8 @@ STATIC void obj_tlssections()
  *      Offset(cseg)         starting offset in cseg
  * Returns:
  *      "segment index" of COMDAT
+ * References:
+ *      COMDAT sections https://www.airs.com/blog/archives/52
  */
 
 STATIC void setup_comdat(Symbol *s)
@@ -1691,11 +1708,73 @@ STATIC void setup_comdat(Symbol *s)
     symbol_debug(s);
     if (tyfunc(s->ty()))
     {
-        //s->Sfl = FLcode;      // was FLoncecode
-        //prefix = ".gnu.linkonce.t";   // doesn't work, despite documentation
+#if !ELF_COMDAT
         prefix = ".text.";              // undocumented, but works
         type = SHT_PROGBITS;
         flags = SHF_ALLOC|SHF_EXECINSTR;
+#else
+        reset_symbuf->write(&s, sizeof(s));
+
+        //s->Sfl = FLcode;      // was FLoncecode
+        /* ".gnu.linkonce.t" sections result in "Too many sections" errors from ld.
+         * prefix = ".gnu.linkonce.t";
+         */
+
+        /* Get section index of the comdat
+         */
+        IDXSEC comdatidx = section_cnt + 1;
+
+        /* create a weak symbol with the group signature in it.
+         *    4  .22 .weak _Z3barIiEiT_,@FUNCTION,VALUE=.text._Z3barIiEiT_+0x00,SIZE=14
+         */
+        const char *p = cpp_mangle(s);
+        IDXSTR namidx = Obj::addstr(symtab_strings, p);
+        IDXSYM info = elf_addsym(namidx, 0, 0, STT_FUNC, STB_WEAK, comdatidx);
+
+        s->Sxtrnnum = info;
+
+        /* Create a section with a type of SHT_GROUP and group flag GRP_COMDAT
+         *    Section 4  .group  GROUP,ENTRIES=2,OFFSET=0x00F0,ALIGN=4,LINK=10,INFO=4
+         *    dd GRP_COMDAT, 5
+         */
+        IDXSTR groupnamidx = section_names->size();
+        section_names->writeString(".group");
+        IDXSTR *pidx = (IDXSTR *)section_names_hashtable->get(&groupnamidx);
+        if (*pidx)
+        {
+            section_names->setsize(groupnamidx);
+            groupnamidx = *pidx;
+        }
+        else
+            *pidx = groupnamidx;
+        IDXSEC groupidx = elf_newsection2(groupnamidx, SHT_GROUP, 0, 0, 0, 0, SHN_SYMTAB, info, 4, 4);
+        int groupseg = elf_getsegment2(groupidx, info, 0);
+        seg_data *pseg = SegData[groupseg];
+        pseg->SDbuf->write32(GRP_COMDAT);
+        pseg->SDbuf->write32(comdatidx);
+
+        /* Create a section with the SHF_GROUP bit set
+         *    Section 5  .text._Z3barIiEiT_  PROGBITS,ALLOC,EXEC,GROUP,SIZE=0x000E(14),OFFSET=0x0060,ALIGN=16
+         */
+        s->Sseg = ElfObj::getsegment(".text.", p, SHT_PROGBITS, SHF_ALLOC|SHF_EXECINSTR|SHF_GROUP, align);
+        IDXSEC comdatidx2 = MAP_SEG2SECIDX(s->Sseg);
+        if (comdatidx2 != comdatidx)
+        {
+            /* This can happen if there are two function definitions with the same signature.
+             * It really should be an error, but it's in the test suite as compile1.d, bug6720()
+             */
+            if (I64)
+                SymbolTable64[STI_FILE].st_shndx = comdatidx2;
+            else
+                SymbolTable[STI_FILE].st_shndx = comdatidx2;
+        }
+
+        if (s->Salignment > align)
+            SegData[s->Sseg]->SDalignment = s->Salignment;
+        SegData[s->Sseg]->SDsym = s;
+        SegData[s->Sseg]->SDassocseg = groupseg;
+        return;
+#endif
     }
     else if ((s->ty() & mTYLINK) == mTYthread)
     {
@@ -1760,7 +1839,69 @@ int Obj::readonly_comdat(Symbol *s)
 
 int Obj::jmpTableSegment(Symbol *s)
 {
-    return (config.flags & CFGromable) ? cseg : CDATA;
+    segidx_t seg = jmpseg;
+    if (seg)                            // memoize the jmpseg on a per-function basis
+        return seg;
+
+    if (config.flags & CFGromable)
+        seg = cseg;
+    else
+    {
+        seg_data *pseg = SegData[s->Sseg];
+        if (pseg->SDassocseg)
+        {
+            /* `s` is in a COMDAT, so the jmp table segment must also
+             * go into its own segment in the same group.
+             */
+            seg = ElfObj::getsegment(".rodata.", s->Sident, SHT_PROGBITS, SHF_ALLOC|SHF_GROUP, NPTRSIZE);
+            addSegmentToComdat(seg, s->Sseg);
+        }
+        else
+            seg = CDATA;
+    }
+    jmpseg = seg;
+    return seg;
+}
+
+/****************************************
+ * If `comdatseg` has a group, add `secidx` to the group.
+ * Params:
+ *      secidx = section to add to the group
+ *      comdatseg = comdat that started the group
+ */
+
+static void addSectionToComdat(IDXSEC secidx, segidx_t comdatseg)
+{
+    seg_data *pseg = SegData[comdatseg];
+    segidx_t groupseg = pseg->SDassocseg;
+    if (groupseg)
+    {
+        seg_data *pgroupseg = SegData[groupseg];
+
+        /* Don't write it if it is already there
+         */
+        Outbuffer *buf = pgroupseg->SDbuf;
+        assert(sizeof(int) == 4);               // loop depends on this
+        for (size_t i = buf->size(); i > 4;)
+        {
+            /* A linear search, but shouldn't be more than 4 items
+             * in it.
+             */
+            i -= 4;
+            if (*(int*)(buf->buf + i) == secidx)
+                return;
+        }
+        buf->write32(secidx);
+    }
+}
+
+/***********************************
+ * Returns:
+ *      jump table segment for function s
+ */
+void addSegmentToComdat(segidx_t seg, segidx_t comdatseg)
+{
+    addSectionToComdat(SegData[seg]->SDshtidx, comdatseg);
 }
 
 /********************************
@@ -2160,8 +2301,12 @@ void Obj::func_start(Symbol *sfunc)
         sfunc->Sseg = CODE;
     //dbg_printf("sfunc->Sseg %d CODE %d cseg %d Coffset %d\n",sfunc->Sseg,CODE,cseg,Offset(cseg));
     cseg = sfunc->Sseg;
+    jmpseg = 0;                         // only 1 jmp seg per function
     assert(cseg == CODE || cseg > COMD);
-    Obj::pubdef(cseg, sfunc, Offset(cseg));
+#if ELF_COMDAT
+    if (!symbol_iscomdat(sfunc))
+#endif
+        Obj::pubdef(cseg, sfunc, Offset(cseg));
     sfunc->Soffset = Offset(cseg);
 
     dwarf_func_start(sfunc);
@@ -2531,6 +2676,7 @@ void ElfObj::addrel(int seg, targ_size_t offset, unsigned type,
 
             relidx = elf_newsection(I64 ? ".rela" : ".rel", p, I64 ? SHT_RELA : SHT_REL, 0);
             segdata->SDrelidx = relidx;
+            addSectionToComdat(relidx,seg);
         }
 
         if (I64)
@@ -2787,11 +2933,15 @@ void Obj::reftodatseg(int seg,targ_size_t offset,targ_size_t val,
         {
             relinfo = R_X86_64_PC32;
             val -= 4;
+            targetsymidx = MAP_SEG2SYMIDX(targetdatum);
         }
         else if (MAP_SEG2SEC(targetdatum)->sh_flags & SHF_TLS)
             relinfo = config.flags3 & CFG3pic ? R_X86_64_TLSGD : R_X86_64_TPOFF32;
         else
-            relinfo = (flags & CFswitch) ? R_X86_64_32S : R_X86_64_32;
+        {
+            relinfo = targetdatum == CDATA ? R_X86_64_32 : R_X86_64_32S;
+            targetsymidx = MAP_SEG2SYMIDX(targetdatum);
+        }
     }
     else
     {
@@ -2801,6 +2951,7 @@ void Obj::reftodatseg(int seg,targ_size_t offset,targ_size_t val,
             relinfo = config.flags3 & CFG3pic ? R_386_TLS_GD : R_386_TLS_LE;
         else
             relinfo = R_386_32;
+        targetsymidx = MAP_SEG2SYMIDX(targetdatum);
     }
     ElfObj::writerel(seg, offset, relinfo, targetsymidx, val);
 }

--- a/src/ddmd/backend/melf.h
+++ b/src/ddmd/backend/melf.h
@@ -406,6 +406,9 @@ typedef struct {
     Elf64_Sxword r_addend;
 } Elf64_Rela;
 
-#define GRP_COMDAT 1
+// Section Group Flags
+#define GRP_COMDAT   1
+#define GRP_MASKOS   0x0ff00000
+#define GRP_MASKPROC 0xf0000000
 
 #endif

--- a/test/runnable/cppa.d
+++ b/test/runnable/cppa.d
@@ -1072,15 +1072,32 @@ void test15579()
     assert(d.y == 8);
 
     printf("d2 = %p\n", d);
-    assert((cast(Interface)d).MethodD() == 3);
+
+    /* Casting to an interface involves thunks in the vtbl[].
+     * g++ puts the thunks for MethodD in the same COMDAT as MethodD.
+     * But D doesn't, so when the linker "picks one" of the D generated MethodD
+     * or the g++ generated MethodD, it may wind up with a messed up thunk,
+     * resulting in a seg fault. The solution is to not expect objects of the same
+     * type to be constructed on both sides of the D/C++ divide if the same member
+     * function (in this case, MethodD) is also defined on both sides.
+     */
+    version (Windows)
+    {
+        assert((cast(Interface)d).MethodD() == 3);
+    }
     assert((cast(Interface)d).MethodCPP() == 30);
+
     assert(d.Method() == 6);
 
     printf("d = %p, i = %p\n", d, cast(Interface)d);
-    Interface i = cppfooi(d);
-    printf("i2: %p\n", i);
-    assert(i.MethodD() == 3);
-    assert(i.MethodCPP() == 30);
+    version (Windows)
+    {
+        Interface i = cppfooi(d);
+        printf("i2: %p\n", i);
+        assert(i.MethodD() == 3);
+        assert(i.MethodCPP() == 30);
+    }
+    printf("test15579() done\n");
 }
 
 /****************************************/

